### ISSUE TITLE
reV version 0.4.17 run on NSRDB v3 sets bad tilt at "latitude"

### ISSUE THREAD

**Issue by EvanRosenliebNREL** [NREL/reV#208]
**Bug Description**
Resource values east of the seam are not matching up with the order of the NSRDB v3 meta. The NSRDB v3 meta has some characteristic unordering, if I were to guess it would appear to be that the resource points are being calculated as straight column major order whereas the meta is not actually ordered that way (for whatever reason)

**Full Traceback**
No error message, just unexpected output

**Code Sample**
reV run in pipeline mode using  /shared-projects/rev/projects/pv_fixtilt_poa/config_gen_test.json. Project points file is in that directory as well

**To Reproduce**
run reV in pipeline mode using version 0.4.17 and config listed above.

**Expected behavior**
I expected the ordering of the resource arrays to be the same as the meta ordering in the output multi-year.h5, however that is not the case. This issue also appears in the single year and chunk files, so issue looks to be in the gen module.

**Screenshots**
Screenshot of a poa variable out of the multiyear file: 

![image](https://user-images.githubusercontent.com/58784383/93119446-10b81380-f67f-11ea-9cd8-d6644e24db0e.png)

Screen shot of cf variable out of one of the single year h5's:

![Screen Shot 2020-09-14 at 11 12 54 AM](https://user-images.githubusercontent.com/58784383/93119615-5e348080-f67f-11ea-9f9e-7b33ad54d873.png)

Screenshot of cf variable out of one of the chunk files:

![Screen Shot 2020-09-14 at 11 16 28 AM](https://user-images.githubusercontent.com/58784383/93119680-760c0480-f67f-11ea-945e-022fd0c3cac5.png)

**System (please complete the following information):**
Run on Linux on eagle
 - Version 0.4.17

**Charge code**
Not sure how this works at end of FY, I will ask Donna if there are hours left in the task...

**Comment by grantbuster** [comment]
Hey @EvanRosenliebNREL, this definitely looks odd but I dont think it's a reV bug. I actually think that this is just a simple plotting effect acting in concert with some weirdness in the NSRDB east vs. west satellite data. Here are some things I think are creating the moire pattern in the east:

1. The NSRDB meta data has bad lat/lon sorting. The west is sorted by one column, the east is sorted by the other. This affects how scatter points are plotted when ordered the same as the meta data or project points. The effect is worsened by scatter point size.
2. The NSRDB east vs. west satellite show different amounts of "noise" in the data. I think this is true of the source data as well as the reV calculated CF. 
3. The mean data definitely compounds these effects. In contrast, take a look at the CF profile outputs plotted for one timestep. The Meta data isn't scrambled. 

Here are some plots to consider with these points:

[test_plots.zip](https://github.com/NREL/reV/files/5220374/test_plots.zip)

**Comment by EvanRosenliebNREL** [comment]
Grant -- 

I do think that looking at it as points as opposed to a raster likely exacerbates the issue, but I don't think it is the source of the issue. 

Here is the mean poa variable for 1998 rasterized: 

![image](https://user-images.githubusercontent.com/58784383/93139410-e590ec80-f69d-11ea-9a52-4d2adb1f54ec.png)

As you can see -- still a lot of "static" east of the seam.

Here's zoomed into somewhere in the central plains around Oklahoma or so:

![image](https://user-images.githubusercontent.com/58784383/93139487-16712180-f69e-11ea-91f8-c3ee22e6e1a8.png)

Meanwhile -- this is what the mean ghi for 1998 looks like at the same extent: 

![image](https://user-images.githubusercontent.com/58784383/93139521-28eb5b00-f69e-11ea-908e-c316ede6a100.png)

As you can see, completely different spatial patterning present. This would seem to imply that the issue is not with differences in the satellite noise between east and west or anything in the NSRDB pipeline as the raw NSRDB variable does not show the same patterning.

An additional interesting point: the GHI variable looks fine whether pulled straight from the NSRDB h5's, or whether pulled from the reV outputs. This seems to imply that rev is passing on the raw NSRDB variables fine but not the SAM variables (the CF variables that Travis looked at earlier also have the same patterning)

So, it seems to be that the divergence between the NSRDB variables and the SAM variables can only be happening either in reV or in SAM.

**Comment by grantbuster** [comment]
I still cant think of a reasonable mechanism that would create these patterns though. If you have a suggested mechanism for the corruption I'm all ears. I don't think that the meta data could be getting shuffled and I don't think that NSRDB resource could be getting mapped incorrectly to the meta data. This plot I sent earlier (along with the CF mean plots) show that there are clear spatial patterns in the reV output. You can see the "noise" but generally the irradiance data looks accurate with sensible cloud patterns. Could there be an ancillary dataset that is corrupt? Maybe, but your images of the POA should be based solely on the irradiance and panel tilt. I'm fairly puzzled! I'll run a few debug tests but I'm not really sure what I'm looking for tbh. 

![image](https://user-images.githubusercontent.com/12700356/93140293-6ef4ee80-f69f-11ea-9e10-859975a2917e.png)

**Comment by EvanRosenliebNREL** [comment]
Grant,
 
I too am puzzled. A corrupted or out of order ancillary dataset seems possible. What would those include?

**Comment by grantbuster** [comment]
PVwatts uses DNI, DHI, air temperature, and wind speed. The air temp and wind speed are sourced from MERRA and are used for a convection calculation within PVwatts that can result in some weirdness sometimes.

**Comment by MRossol** [comment]
Well I'm glad to see the irradiance isn't f'd up! The variables used for CF are irradiance + wind speed + temperature. Evan, can you quickly map of mean Temperature and Wind_Speed to see if either of those have the same pattern?

Another CRAZY thought is could this be an artifact of the low precision of the NSRDB lat, lon coordinates since this is fixed tilt at lat? An easy test would be to re-run the analysis with a set tilt in all locations...

**Comment by grantbuster** [comment]
Yeah @MRossol or run a 1-axis tracking test. I'll do that now

**Comment by EvanRosenliebNREL** [comment]
Grand And Rossol,

Something that is really, really hard to figure out for both of these scenarios (something to do with it being fixed tilt, or something to do with the MERRA joining) is why is it only a problem east of the seam?

The correlation between poa and ghi west of the seam is *extremely* high. Just to choose a random example, here is the snake river valley in Idaho.

poa:

![image](https://user-images.githubusercontent.com/58784383/93141497-83d28180-f6a1-11ea-8b30-687c78f9906c.png)

ghi:

![image](https://user-images.githubusercontent.com/58784383/93141518-8d5be980-f6a1-11ea-8a7c-4c6ff40880f7.png)

**Comment by MRossol** [comment]
What does DNI look like in that location?

**Comment by EvanRosenliebNREL** [comment]
Short story: Air temp, windspeed, DNI and DHI don't appear to help explain the mystery.

Many pictures incoming, apologies. Not going to post the windspeed pictures: the resource is from such an obviously lower resolution that it is clear that it is not causing this particular pattern.

For the mid plains region, here is a picture of POA again just because the extents don't exactly line up:

![image](https://user-images.githubusercontent.com/58784383/93167231-6b805800-f6dd-11ea-95a4-7705995d7537.png)

DHI:

![image](https://user-images.githubusercontent.com/58784383/93167250-789d4700-f6dd-11ea-9ecc-38cd23616dfb.png)

DNI:

![image](https://user-images.githubusercontent.com/58784383/93167263-84890900-f6dd-11ea-998d-54e67987f765.png)

Air Temp:

![image](https://user-images.githubusercontent.com/58784383/93167295-94a0e880-f6dd-11ea-93a0-46e2fcb8d37f.png)

None showing anything like what is seen in the POA variable.

For the Snake River Valley Region:

POA:

![image](https://user-images.githubusercontent.com/58784383/93167455-e77aa000-f6dd-11ea-904e-23e8f54ab962.png)

DHI:

![image](https://user-images.githubusercontent.com/58784383/93167499-fcefca00-f6dd-11ea-93ea-c55a45d5b112.png)

DNI:

![image](https://user-images.githubusercontent.com/58784383/93167526-05480500-f6de-11ea-9448-8b6e88358ea9.png)

Temp:

![image](https://user-images.githubusercontent.com/58784383/93167552-109b3080-f6de-11ea-9613-23403b696b27.png)

Full CONUS

DHI:

![image](https://user-images.githubusercontent.com/58784383/93167588-26a8f100-f6de-11ea-9176-2e5ce84c4ddc.png)

DNI:

![image](https://user-images.githubusercontent.com/58784383/93167608-31fc1c80-f6de-11ea-83ab-395542f3dd45.png)

Air Temp:

![image](https://user-images.githubusercontent.com/58784383/93167620-3b858480-f6de-11ea-8dda-99d9be05c686.png)

**Comment by MRossol** [comment]
I think this is just a satellite artifact/point ordering artifact and not an actual bug.

**Comment by grantbuster** [comment]
https://github.com/NREL/reV/pull/209

**Comment by grantbuster** [comment]
Bug summary and closure: 

Hey all,

We found a real bug in reV. 

Huge shout out to Evan for pushing the issue, thinking critically, and not giving in to superficial explanations. 

When a user set a PV panel to tilt @ latitude, the SAM config dictionary is modified to set the tilt equal to the actual latitude float value. The problem was that a pointer to the SAM config object was shared between workers on a serial compute chunk, so once the first modification was made, all downstream workers inherited the first tilt value. So if we were running with 100 sites per worker (a typical serial compute chunk), all 100 sites would have the same tilt. This is fairly accurate until you look at the ordering of the NSRDB meta data in the eastern extent (see screenshot below). The latitudes are totally scrambled so a group of 100 consecutive sites might have latitudes with a massive range. Ironically, if the NSRDB meta data hadn’t been so messed up, we may never have found this bug! 

The bottom line is that this bug adds “noise” to the capacity factor and plane-of-array irradiance in the eastern sites (see below and attached, the bad data looks kind of “grainy”). The irradiance variables are being passed through correctly but the tilt is off by a few degrees. I’m happy to do a more rigorous delta-study if that’s important to understand. I’m not sure what the research implications of this are, so feel free to suggest corrective actions. 

Details and plots:
[reV_pv_tilt_bug.zip](https://github.com/NREL/reV/files/5233008/reV_pv_tilt_bug.zip)
